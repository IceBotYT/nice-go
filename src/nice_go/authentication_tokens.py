""" " Holds the tokens retrieved from authentication."""

from __future__ import annotations


class AuthenticationTokens:
    def __init__(self, data: dict[str, str]) -> None:
        self.id_token = data["IdToken"]
        self.refresh_token: str | None = None
        try:
            self.refresh_token = data["RefreshToken"]
        except KeyError:
            self.refresh_token = None
