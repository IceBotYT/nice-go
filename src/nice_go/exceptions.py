class NiceGOError(Exception):
    """Base exception for Nice G.O. API"""


class NoAuthError(NiceGOError):
    """Not authenticated exception."""


class ApiError(NiceGOError):
    """Api error."""


class AuthFailedError(NiceGOError):
    """Authentication failed."""


class WebSocketError(NiceGOError):
    """Websocket error."""
