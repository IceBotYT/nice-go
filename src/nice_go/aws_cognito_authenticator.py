"""AWS Cognito authentication and identity management."""

import logging
from concurrent.futures import ThreadPoolExecutor

import boto3
from aiobotocore.session import get_session
from pycognito import AWSSRP  # type: ignore[import-untyped]

from nice_go.authentication_tokens import AuthenticationTokens

_LOGGER = logging.getLogger(__name__)

executor = ThreadPoolExecutor(max_workers=5)


class AwsCognitoAuthenticator:
    def __init__(
        self,
        region_name: str,
        client_id: str,
        pool_id: str,
        identity_pool_id: str,
    ) -> None:
        self.region_name = region_name
        self.identity_pool_id = identity_pool_id
        self.client_id = client_id
        self.pool_id = pool_id
        self.session = get_session()

    """ Regenerates the token by providing a refresh token. """

    def refresh_token(self, refresh_token: str) -> AuthenticationTokens:
        cognito_identity_provider = boto3.client("cognito-idp", self.region_name)
        resp = cognito_identity_provider.initiate_auth(
            AuthFlow="REFRESH_TOKEN_AUTH",
            AuthParameters={
                "REFRESH_TOKEN": refresh_token,
            },
            ClientId=self.client_id,
        )
        _LOGGER.debug("Authentication response %s", resp)
        return AuthenticationTokens(resp["AuthenticationResult"])

    """ Gets the initial token by providing username and password. """

    def get_new_token(self, username: str, password: str) -> AuthenticationTokens:
        cognito_identity_provider = boto3.client("cognito-idp", self.region_name)
        # Start the authentication flow
        aws_srp = AWSSRP(
            username=username,
            password=password,
            pool_id=self.pool_id,
            client_id=self.client_id,
            client=cognito_identity_provider,
        )

        auth_params = aws_srp.get_auth_params()
        resp = cognito_identity_provider.initiate_auth(
            AuthFlow="USER_SRP_AUTH",
            AuthParameters=auth_params,
            ClientId=self.client_id,
        )

        challenge_response = aws_srp.process_challenge(
            resp["ChallengeParameters"],
            auth_params,
        )

        # Respond to PASSWORD_VERIFIER
        resp = cognito_identity_provider.respond_to_auth_challenge(
            ClientId=self.client_id,
            ChallengeName="PASSWORD_VERIFIER",
            ChallengeResponses=challenge_response,
        )
        _LOGGER.debug("Authentication result %s", resp)
        return AuthenticationTokens(resp["AuthenticationResult"])
