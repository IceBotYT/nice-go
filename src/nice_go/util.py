from __future__ import annotations

import json
from typing import Any

from nice_go.const import REQUEST_TEMPLATES


async def get_request_template(
    request_name: str,
    arguments: dict[str, str] | None,
) -> Any:
    template = json.dumps(REQUEST_TEMPLATES[request_name])
    if arguments:
        for key, value in arguments.items():
            template = template.replace(f"${key}", value)
    return json.loads(template)
