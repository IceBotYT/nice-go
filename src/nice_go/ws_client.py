from __future__ import annotations

import asyncio
import base64
import json
import uuid
from typing import TYPE_CHECKING, Any, Callable, NamedTuple

import aiohttp

from nice_go.exceptions import WebSocketError
from nice_go.util import get_request_template

if TYPE_CHECKING:
    import yarl


class EventListener(NamedTuple):
    predicate: Callable[[dict[str, Any]], bool] | None
    event: str
    result: Callable[[dict[str, Any]], Any] | None
    future: asyncio.Future[Any]


class WebSocketClient:
    def __init__(self) -> None:
        self.ws: aiohttp.ClientWebSocketResponse | None = None
        self._dispatch_listeners: list[EventListener] = []
        self._subscriptions: list[str] = []

    async def connect(
        self,
        client_session: aiohttp.ClientSession,
        id_token: str,
        endpoint: yarl.URL,
        dispatch: Callable[[str, dict[str, Any] | None], None],
        host: str | None = None,
    ) -> None:
        if host is None:
            msg = "host must be provided"
            raise ValueError(msg)
        self._dispatch = dispatch
        self.id_token = id_token
        self.host = host

        header_dict = {
            "Authorization": id_token,
            "host": host,
        }
        # Base64 encode the header
        header = base64.b64encode(json.dumps(header_dict).encode()).decode()
        # Construct the URL
        url = endpoint.with_query({"header": header, "payload": "e30="})

        headers = {"sec-websocket-protocol": "graphql-ws"}
        self.ws = await client_session.ws_connect(url, headers=headers)

        await self.init()

    async def init(self) -> None:
        if self.ws is None or self.ws.closed:
            msg = "WebSocket connection is closed"
            raise WebSocketError(msg)
        await self.send({"type": "connection_init"})
        try:
            message = await self.ws.receive(timeout=10)
            data = json.loads(message.data)
            if data["type"] != "connection_ack":
                msg = "Expected connection_ack, but received {}".format(data["type"])
                raise WebSocketError(
                    msg,
                )
        except asyncio.TimeoutError as e:
            msg = "Connection to the websocket server timed out"
            raise WebSocketError(msg) from e

        self._dispatch("connected", None)

    async def send(self, message: str | dict[str, Any]) -> None:
        if self.ws is None or self.ws.closed:
            msg = "WebSocket connection is closed"
            raise WebSocketError(msg)
        if isinstance(message, dict):
            await self.ws.send_json(message)
        else:
            await self.ws.send_str(message)

    async def close(self) -> None:
        if self.ws is None or self.ws.closed:
            return
        # Unsubscribe from all subscriptions
        for subscription_id in self._subscriptions:
            await self.unsubscribe(subscription_id)
        await self.ws.close()

    async def poll(self) -> None:
        if self.ws is None or self.ws.closed:
            error_msg = "WebSocket connection is closed"
            raise WebSocketError(error_msg)
        msg = await self.ws.receive(timeout=60.0)
        if msg.type == aiohttp.WSMsgType.TEXT:
            await self.received_message(msg.data)
        elif msg.type == aiohttp.WSMsgType.ERROR:
            error_msg = f"WebSocket connection closed with error {msg}"
            raise WebSocketError(error_msg)
        elif msg.type in (
            aiohttp.WSMsgType.CLOSE,
            aiohttp.WSMsgType.CLOSING,
            aiohttp.WSMsgType.CLOSED,
        ):
            error_msg = "WebSocket connection closed"
            raise WebSocketError(error_msg)

    def load_message(self, message: str) -> Any:
        try:
            parsed_message = json.loads(message)
        except json.JSONDecodeError as e:
            msg = f"Received invalid JSON message: {message}"
            raise WebSocketError(msg) from e

        return parsed_message

    def dispatch_message(self, message: dict[str, Any]) -> None:
        if message["type"] == "data":
            self._dispatch(message["type"], message["payload"])
        elif message["type"] == "error":
            msg = f"Received error message: {message}"
            raise WebSocketError(msg)

    async def received_message(self, message: str) -> None:
        parsed_message = self.load_message(message)
        self.dispatch_message(parsed_message)

        removed = []
        for index, entry in enumerate(self._dispatch_listeners):
            if entry.event != parsed_message["type"]:
                continue

            future = entry.future
            if future.cancelled():
                removed.append(index)
                continue

            if entry.predicate is not None:
                try:
                    valid = entry.predicate(parsed_message)
                except Exception as e:  # noqa: BLE001
                    future.set_exception(e)
                    removed.append(index)
                    continue
            else:
                valid = True

            if valid:
                ret = (
                    parsed_message
                    if entry.result is None
                    else entry.result(parsed_message)
                )
                future.set_result(ret)
                removed.append(index)

        for index in reversed(removed):
            del self._dispatch_listeners[index]

    def wait_for(
        self,
        event: str,
        predicate: Callable[[dict[str, Any]], bool] | None = None,
        result: Callable[[dict[str, Any]], Any] | None = None,
    ) -> asyncio.Future[Any]:
        future: asyncio.Future[dict[str, Any]] = asyncio.Future()
        self._dispatch_listeners.append(EventListener(predicate, event, result, future))
        return future

    async def subscribe(self, receiver: str) -> str:
        subscription_id = str(uuid.uuid4())
        payload = await get_request_template(
            "subscribe",
            {
                "receiver_id": receiver,
                "uuid": subscription_id,
                "id_token": self.id_token,
                "host": self.host,
            },
        )
        await self.send(payload)

        def _predicate(message: dict[str, Any]) -> bool:
            if "type" not in message:
                return False
            if "id" not in message:
                return False
            valid: bool = (
                message["type"] == "start_ack" and message["id"] == subscription_id
            )
            return valid

        try:
            await asyncio.wait_for(self.wait_for("start_ack", _predicate), timeout=10)
        except asyncio.TimeoutError as e:
            msg = "Subscription to the websocket server timed out"
            raise WebSocketError(msg) from e

        self._subscriptions.append(subscription_id)

        return subscription_id

    async def unsubscribe(self, subscription_id: str) -> None:
        payload = await get_request_template("unsubscribe", {"id": subscription_id})
        await self.send(payload)
        self._subscriptions.remove(subscription_id)

    @property
    def closed(self) -> bool:
        if self.ws is None:
            return True
        return self.ws.closed
