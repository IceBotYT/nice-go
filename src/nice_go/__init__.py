from nice_go.barrier import Barrier
from nice_go.const import BARRIER_STATUS
from nice_go.exceptions import (
    ApiError,
    AuthFailedError,
    NiceGOError,
    NoAuthError,
    WebSocketError,
)
from nice_go.nice_go_api import NiceGOApi

__all__ = [
    "BARRIER_STATUS",
    "Barrier",
    "NiceGOApi",
    "ApiError",
    "AuthFailedError",
    "NiceGOError",
    "WebSocketError",
    "NoAuthError",
]
