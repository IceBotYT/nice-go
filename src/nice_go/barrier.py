# ruff: noqa: SLF001

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    # This is a forward reference to avoid circular imports
    from datetime import datetime

    from nice_go.nice_go_api import NiceGOApi


@dataclass
class ConnectionState:
    """Class representing the connection state of a barrier."""

    connected: bool
    updatedTimestamp: datetime  # noqa: N815


@dataclass
class BarrierState:
    """Class representing the state of a barrier."""

    deviceId: str  # noqa: N815
    desired: dict[str, Any]
    reported: dict[str, Any]
    timestamp: str
    version: str
    connectionState: ConnectionState | None  # noqa: N815


@dataclass
class Barrier:
    """Class representing a barrier."""

    id: str
    type: str
    controlLevel: str  # noqa: N815
    attr: list[dict[str, str]]
    state: BarrierState
    api: NiceGOApi

    async def open(self) -> bool:
        """Open the barrier."""
        return await self.api._open_barrier(self.id)

    async def close(self) -> bool:
        """Close the barrier."""
        return await self.api._close_barrier(self.id)

    async def light_on(self) -> bool:
        """Turn on the light of the barrier."""
        return await self.api._light_on(self.id)

    async def light_off(self) -> bool:
        """Turn off the light of the barrier."""
        return await self.api._light_off(self.id)

    async def get_attr(self, key: str) -> str:
        """Get the value of an attribute."""
        attr = next((attr for attr in self.attr if attr["key"] == key), None)
        if attr is None:
            msg = f"Attribute with key {key} not found."
            raise ValueError(msg)
        return attr["value"]
