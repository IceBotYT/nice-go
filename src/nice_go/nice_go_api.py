"""Parses data from the Nice G.O. API."""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Callable, Coroutine, TypeVar

import aiohttp
import botocore
import yarl

from nice_go.aws_cognito_authenticator import AwsCognitoAuthenticator
from nice_go.backoff import ExponentialBackoff
from nice_go.barrier import Barrier, BarrierState, ConnectionState
from nice_go.const import (
    CLIENT_ID,
    ENDPOINTS_URL,
    IDENTITY_POOL_ID,
    POOL_ID,
    REGION_NAME,
)
from nice_go.exceptions import (
    ApiError,
    AuthFailedError,
    NoAuthError,
    WebSocketError,
)
from nice_go.util import get_request_template
from nice_go.ws_client import WebSocketClient

T = TypeVar("T")
Coro = Coroutine[Any, Any, T]
CoroT = TypeVar("CoroT", bound=Callable[..., Coro[Any]])

_LOGGER = logging.getLogger(__name__)


class NiceGOApi:
    def __init__(self) -> None:
        self.id_token: str | None = None
        self._closing_task: asyncio.Task[None] | None = None
        self.ws: WebSocketClient | None = None
        self.endpoints: dict[str, Any] | None = None
        self.session: aiohttp.ClientSession | None = None
        self._event_tasks: set[asyncio.Task[None]] = set()

        self.authenticator = AwsCognitoAuthenticator(
            REGION_NAME,
            CLIENT_ID,
            POOL_ID,
            IDENTITY_POOL_ID,
        )

    def event(self, coro: CoroT) -> CoroT:
        """Decorator to add an event listener."""
        if not asyncio.iscoroutinefunction(coro):
            msg = "The decorated function must be a coroutine"
            raise TypeError(msg)

        setattr(self, coro.__name__, coro)
        return coro

    async def _run_event(
        self,
        coro: Callable[..., Coroutine[Any, Any, Any]],
        event_name: str,
        data: dict[str, Any] | None = None,
    ) -> None:
        kwargs = {}
        if data is not None:
            kwargs["data"] = data
        try:
            await coro(**kwargs)
        except asyncio.CancelledError:
            pass
        except Exception:
            _LOGGER.exception("Error while handling event %s", event_name)

    def _schedule_event(
        self,
        coro: Callable[..., Coroutine[Any, Any, Any]],
        event_name: str,
        data: dict[str, Any] | None,
    ) -> None:
        """Schedule an event to be dispatched."""
        wrapped = self._run_event(coro, event_name, data)
        task = asyncio.create_task(wrapped, name=f"NiceGO: {event_name}")
        self._event_tasks.add(task)  # See RUF006
        task.add_done_callback(self._event_tasks.discard)

    def dispatch(self, event: str, data: dict[str, Any] | None = None) -> None:
        """Dispatch an event to the listeners."""
        method = "on_" + event

        try:
            coro = getattr(self, method)
        except AttributeError:
            pass
        else:
            self._schedule_event(coro, method, data)

    async def authenticate_refresh(
        self,
        refresh_token: str,
        session: aiohttp.ClientSession,
    ) -> str | None:
        """Authenticate using a previous obtained refresh token."""
        self.session = session
        return await self.__authenticate(None, None, refresh_token)

    async def authenticate(
        self,
        user_name: str,
        password: str,
        session: aiohttp.ClientSession,
    ) -> str | None:
        """Authenticate using username and password."""
        self.session = session
        return await self.__authenticate(user_name, password, None)

    async def __authenticate(
        self,
        user_name: str | None,
        password: str | None,
        refresh_token: str | None,
    ) -> str | None:
        try:
            if refresh_token is None:
                if user_name is None or password is None:
                    msg = "Username and password must be provided"
                    raise ValueError(msg)
                token = await asyncio.to_thread(
                    self.authenticator.get_new_token,
                    user_name,
                    password,
                )
            else:
                token = await asyncio.to_thread(
                    self.authenticator.refresh_token,
                    refresh_token,
                )

            self.id_token = token.id_token

            if self.session is None:
                msg = "ClientSession not provided"
                raise ValueError(msg)

            # Get the endpoints while we're at it
            data = await self.session.get(ENDPOINTS_URL)
            endpoints = await data.json()
            self.endpoints = endpoints["endpoints"]
        except botocore.exceptions.ClientError as e:
            _LOGGER.exception("Exception while authenticating")
            if e.response["Error"]["Code"] == "NotAuthorizedException":
                raise AuthFailedError from e
            raise ApiError from e
        else:
            return token.refresh_token

    @property
    def closed(self) -> bool:
        return self._closing_task is not None

    async def connect(self, *, reconnect: bool = True) -> None:
        backoff = ExponentialBackoff()
        try:
            while not self.closed:
                if self.id_token is None:
                    raise NoAuthError

                if self.endpoints is None:
                    msg = "Endpoints not available"
                    raise ApiError(msg)

                if self.session is None:
                    msg = "ClientSession not provided"
                    raise ValueError(msg)

                api_url = self.endpoints["GraphQL"]["device"]["wss"]

                self.ws = WebSocketClient()
                await self.ws.connect(
                    self.session,
                    self.id_token,
                    yarl.URL(api_url),
                    self.dispatch,
                    yarl.URL(self.endpoints["GraphQL"]["device"]["https"]).host,
                )

                while True:
                    await self.ws.poll()
        except (
            OSError,
            WebSocketError,
            aiohttp.ClientError,
            asyncio.TimeoutError,
        ) as e:
            self.dispatch("connection_lost", {"exception": e})
            if not reconnect:
                await self.close()
                raise

            if self.closed:
                return

            retry = backoff.delay()
            _LOGGER.exception("Connection lost, retrying in %s seconds", retry)
            await asyncio.sleep(retry)

    async def subscribe(self, receiver: str) -> str:
        if self.ws is None:
            msg = "No WebSocket connection"
            raise WebSocketError(msg)

        return await self.ws.subscribe(receiver)

    async def unsubscribe(self, receiver: str) -> None:
        if self.ws is None:
            msg = "No WebSocket connection"
            raise WebSocketError(msg)

        await self.ws.unsubscribe(receiver)

    async def close(self) -> None:
        async def _close() -> None:
            if self.ws:
                await self.ws.close()

        self._closing_task = asyncio.create_task(_close())
        await self._closing_task

    async def get_all_barriers(self) -> list[Barrier]:
        if self.id_token is None:
            raise NoAuthError
        if self.session is None:
            msg = "ClientSession not provided"
            raise ValueError(msg)
        if self.endpoints is None:
            msg = "Endpoints not available"
            raise ApiError(msg)

        api_url = self.endpoints["GraphQL"]["device"]["https"]

        headers = {"Authorization": self.id_token, "Content-Type": "application/json"}

        response = await self.session.post(
            api_url,
            headers=headers,
            json=await get_request_template("get_all_barriers", None),
        )
        data = await response.json()

        barriers = []

        for device in data["data"]["devicesListAll"]["devices"]:
            if device["state"]["connectionState"] is not None:
                connection_state = ConnectionState(
                    device["state"]["connectionState"]["connected"],
                    device["state"]["connectionState"]["updatedTimestamp"],
                )
            else:
                connection_state = None
            barrier_state = BarrierState(
                device["state"]["deviceId"],
                json.loads(device["state"]["desired"]),
                json.loads(device["state"]["reported"]),
                device["state"]["timestamp"],
                device["state"]["version"],
                connection_state,
            )
            barrier = Barrier(
                device["id"],
                device["type"],
                device["controlLevel"],
                device["attr"],
                barrier_state,
                self,
            )
            barriers.append(barrier)

        return barriers

    async def _open_barrier(self, barrier_id: str) -> bool:
        if self.id_token is None:
            raise NoAuthError
        if self.session is None:
            msg = "ClientSession not provided"
            raise ValueError(msg)
        if self.endpoints is None:
            msg = "Endpoints not available"
            raise ApiError(msg)

        api_url = self.endpoints["GraphQL"]["device"]["https"]

        headers = {"Authorization": self.id_token, "Content-Type": "application/json"}

        response = await self.session.post(
            api_url,
            headers=headers,
            json=await get_request_template("open_barrier", {"barrier_id": barrier_id}),
        )
        data = await response.json()
        result: bool = data["data"]["devicesControl"]

        return result

    async def _close_barrier(self, barrier_id: str) -> bool:
        if self.id_token is None:
            raise NoAuthError
        if self.session is None:
            msg = "ClientSession not provided"
            raise ValueError(msg)
        if self.endpoints is None:
            msg = "Endpoints not available"
            raise ApiError(msg)

        api_url = self.endpoints["GraphQL"]["device"]["https"]

        headers = {"Authorization": self.id_token, "Content-Type": "application/json"}

        response = await self.session.post(
            api_url,
            headers=headers,
            json=await get_request_template(
                "close_barrier",
                {"barrier_id": barrier_id},
            ),
        )
        data = await response.json()
        result: bool = data["data"]["devicesControl"]

        return result

    async def _light_on(self, barrier_id: str) -> bool:
        if self.id_token is None:
            raise NoAuthError
        if self.session is None:
            msg = "ClientSession not provided"
            raise ValueError(msg)
        if self.endpoints is None:
            msg = "Endpoints not available"
            raise ApiError(msg)

        api_url = self.endpoints["GraphQL"]["device"]["https"]

        headers = {"Authorization": self.id_token, "Content-Type": "application/json"}

        response = await self.session.post(
            api_url,
            headers=headers,
            json=await get_request_template("light_on", {"barrier_id": barrier_id}),
        )
        data = await response.json()
        result: bool = data["data"]["devicesControl"]

        return result

    async def _light_off(self, barrier_id: str) -> bool:
        if self.id_token is None:
            raise NoAuthError
        if self.session is None:
            msg = "ClientSession not provided"
            raise ValueError(msg)
        if self.endpoints is None:
            msg = "Endpoints not available"
            raise ApiError(msg)

        api_url = self.endpoints["GraphQL"]["device"]["https"]

        headers = {"Authorization": self.id_token, "Content-Type": "application/json"}

        response = await self.session.post(
            api_url,
            headers=headers,
            json=await get_request_template("light_off", {"barrier_id": barrier_id}),
        )
        data = await response.json()
        result: bool = data["data"]["devicesControl"]

        return result
