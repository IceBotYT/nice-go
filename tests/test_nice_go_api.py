# ruff: noqa: SLF001
from __future__ import annotations

import asyncio
from contextlib import suppress
from typing import TYPE_CHECKING, Any
from unittest.mock import AsyncMock, MagicMock, patch

import botocore
import pytest
from nice_go import (
    ApiError,
    AuthFailedError,
    NiceGOApi,
    NoAuthError,
    WebSocketError,
)

from tests.const import (
    GET_ALL_BARRIERS_RESPONSE,
    GET_ALL_BARRIERS_RESPONSE_NO_CONNECTION_STATE,
)

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion


async def test_schedule_event(mock_api: NiceGOApi) -> None:
    with patch("nice_go.nice_go_api.asyncio.create_task") as mock_create_task:
        coro = AsyncMock()
        mock_api._schedule_event(coro, "test_event", {"key": "value"})
        mock_create_task.assert_called_once()
        await mock_create_task.call_args[0][0]  # Await the coroutine that was scheduled


async def test_dispatch_event(mock_api: NiceGOApi) -> None:
    coro = AsyncMock()
    mock_api.on_test_event = coro  # type: ignore[attr-defined]
    mock_api.dispatch("test_event", {"key": "value"})
    await asyncio.sleep(0)  # Allow the event loop to run
    coro.assert_called_once()


async def test_authenticate(mock_api: NiceGOApi) -> None:
    with patch("nice_go.nice_go_api.asyncio.to_thread") as mock_to_thread:
        mock_to_thread.return_value = MagicMock(
            id_token="test_token",  # noqa: S106
            refresh_token="refresh_token",  # noqa: S106
        )
        assert mock_api.session is not None
        result = await mock_api.authenticate("username", "password", mock_api.session)
        assert result == "refresh_token"
        assert mock_api.id_token == "test_token"


async def test_connect_not_authenticated(mock_api: NiceGOApi) -> None:
    mock_api.id_token = None
    with pytest.raises(NoAuthError):
        await mock_api.connect()


async def test_connect_no_endpoints(mock_api: NiceGOApi) -> None:
    mock_api.id_token = "test_token"
    mock_api.endpoints = None
    with pytest.raises(ApiError):
        await mock_api.connect()


async def test_subscribe_no_ws(mock_api: NiceGOApi) -> None:
    mock_api.ws = None
    with pytest.raises(WebSocketError):
        await mock_api.subscribe("receiver")


async def test_unsubscribe_no_ws(mock_api: NiceGOApi) -> None:
    mock_api.ws = None
    with pytest.raises(WebSocketError):
        await mock_api.unsubscribe("receiver")


async def test_get_all_barriers_not_authenticated(mock_api: NiceGOApi) -> None:
    mock_api.id_token = None
    with pytest.raises(NoAuthError):
        await mock_api.get_all_barriers()


@pytest.mark.parametrize(
    ("method_name", "query"),
    [
        ("_open_barrier", "open_barrier"),
        ("_close_barrier", "close_barrier"),
        ("_light_on", "light_on"),
        ("_light_off", "light_off"),
    ],
)
async def test_barrier_operations(
    mock_api: NiceGOApi,
    method_name: str,
    query: str,
) -> None:
    with patch("nice_go.nice_go_api.get_request_template") as mock_get_request_template:
        mock_api.id_token = "test_token"
        mock_get_request_template.return_value = {"query": query}
        assert mock_api.session is not None
        assert isinstance(mock_api.session, MagicMock)
        mock_api.session.post.return_value.json.return_value = {
            "data": {"devicesControl": True},
        }
        method = getattr(mock_api, method_name)
        result = await method("barrier_id")
        assert result is True


async def test_event_decorator(mock_api: NiceGOApi) -> None:
    @mock_api.event
    async def on_test_event(data: dict[str, Any]) -> None:
        pass

    assert "on_test_event" in dir(mock_api)
    assert mock_api.on_test_event == on_test_event  # type: ignore[attr-defined]
    mock_api.dispatch("test_event", {})


async def test_sync_event_decorator(mock_api: NiceGOApi) -> None:
    with pytest.raises(TypeError):

        @mock_api.event  # type: ignore[type-var]
        def on_test_event(data: dict[str, Any]) -> None:
            # It's impossible for this to be called because
            # the error stops it from being added
            pass  # pragma: no cover


@pytest.mark.parametrize("error", [asyncio.CancelledError, Exception])
async def test_run_event_errors(mock_api: NiceGOApi, error: Exception) -> None:
    coro = AsyncMock()
    coro.side_effect = error
    mock_api.on_test_event = coro  # type: ignore[attr-defined]
    mock_api.dispatch("test_event", {})
    await asyncio.sleep(0)  # Allow the event loop to run
    coro.assert_called_once()


async def test_dispatch_no_listener(mock_api: NiceGOApi) -> None:
    mock_api.dispatch("test_event", {})
    await asyncio.sleep(0)  # Allow the event loop to run


async def test_authenticate_refresh(mock_api: NiceGOApi) -> None:
    with patch("nice_go.nice_go_api.asyncio.to_thread") as mock_to_thread:
        mock_to_thread.return_value = MagicMock(
            id_token="test_token",  # noqa: S106
            refresh_token="refresh_token",  # noqa: S106
        )
        mock_api.id_token = "test_token"
        assert mock_api.session is not None
        result = await mock_api.authenticate_refresh(
            "refresh_token",
            session=mock_api.session,
        )
        assert result == "refresh_token"


async def test_authenticate_botocore_client_error(mock_api: NiceGOApi) -> None:
    with patch("nice_go.nice_go_api.asyncio.to_thread") as mock_to_thread:
        mock_to_thread.side_effect = botocore.exceptions.ClientError(
            error_response={"Error": {"Code": "TestException"}},
            operation_name="test",
        )
        assert mock_api.session is not None
        with pytest.raises(ApiError):
            await mock_api.authenticate("username", "password", mock_api.session)


async def test_authenticate_not_authorized_exception(mock_api: NiceGOApi) -> None:
    with patch("nice_go.nice_go_api.asyncio.to_thread") as mock_to_thread:
        mock_to_thread.side_effect = botocore.exceptions.ClientError(
            error_response={"Error": {"Code": "NotAuthorizedException"}},
            operation_name="test",
        )
        assert mock_api.session is not None
        with pytest.raises(AuthFailedError):
            await mock_api.authenticate("username", "password", mock_api.session)


async def test_connect_error(mock_api: NiceGOApi) -> None:
    mock_api.id_token = "test_token"

    with patch("nice_go.nice_go_api.WebSocketClient") as mock_ws_client:
        mock_ws_client_instance = AsyncMock()
        mock_ws_client.return_value = mock_ws_client_instance
        mock_ws_client_instance.poll.side_effect = WebSocketError()
        with pytest.raises(WebSocketError):
            await mock_api.connect(reconnect=False)
        mock_ws_client_instance.connect.assert_called_once()


async def test_connect_closed(mock_api: NiceGOApi) -> None:
    mock_api.id_token = "test_token"

    with patch("nice_go.nice_go_api.WebSocketClient") as mock_ws_client:
        mock_ws_client_instance = AsyncMock()
        mock_ws_client.return_value = mock_ws_client_instance

        async def side_effect() -> None:
            mock_api._closing_task = asyncio.create_task(asyncio.sleep(0))
            await asyncio.sleep(0)
            raise WebSocketError

        mock_ws_client_instance.poll.side_effect = side_effect
        await mock_api.connect(reconnect=True)
        mock_ws_client_instance.connect.assert_called_once()


async def test_connect_reconnect(mock_api: NiceGOApi) -> None:
    mock_api.id_token = "test_token"

    with patch("nice_go.nice_go_api.WebSocketClient") as mock_ws_client:
        mock_ws_client_instance = AsyncMock()
        mock_ws_client.return_value = mock_ws_client_instance
        mock_ws_client_instance.poll.side_effect = [WebSocketError(), None]
        with suppress(StopAsyncIteration):
            await mock_api.connect(reconnect=True)
        connect_call_count = 2
        assert mock_ws_client_instance.connect.call_count == connect_call_count


async def test_subscribe(mock_api: NiceGOApi) -> None:
    mock_api.id_token = "test_token"
    mock_api.ws = AsyncMock()
    mock_api.ws.subscribe.return_value = "test_id"
    result = await mock_api.subscribe("receiver")
    assert result == "test_id"


async def test_unsubscribe(mock_api: NiceGOApi) -> None:
    mock_api.id_token = "test_token"
    mock_api.ws = AsyncMock()
    await mock_api.unsubscribe("test_id")
    mock_api.ws.unsubscribe.assert_called_once_with("test_id")


async def test_get_all_barriers(
    mock_api: NiceGOApi,
    snapshot: SnapshotAssertion,
) -> None:
    mock_api.id_token = "test_token"
    assert mock_api.session is not None
    assert isinstance(mock_api.session, MagicMock)
    mock_api.session.post.return_value.json.return_value = GET_ALL_BARRIERS_RESPONSE
    result = await mock_api.get_all_barriers()
    # api is an object with an address that varies, so we exclude it from the snapshot
    # It's not what we're checking anyways
    # Remove the api property from all the barriers
    for barrier in result:
        barrier.api = None  # type: ignore[assignment]
    assert result == snapshot


async def test_get_all_barriers_no_connection_state(
    mock_api: NiceGOApi,
    snapshot: SnapshotAssertion,
) -> None:
    mock_api.id_token = "test_token"
    assert mock_api.session is not None
    assert isinstance(mock_api.session, MagicMock)
    mock_api.session.post.return_value.json.return_value = (
        GET_ALL_BARRIERS_RESPONSE_NO_CONNECTION_STATE
    )
    result = await mock_api.get_all_barriers()
    for barrier in result:
        barrier.api = None  # type: ignore[assignment]

    assert result == snapshot


@pytest.mark.parametrize(
    ("method_name"),
    [
        ("_open_barrier"),
        ("_close_barrier"),
        ("_light_on"),
        ("_light_off"),
    ],
)
async def test_barrier_operations_no_auth(
    mock_api: NiceGOApi,
    method_name: str,
) -> None:
    mock_api.id_token = None
    method = getattr(mock_api, method_name)
    with pytest.raises(NoAuthError):
        await method("barrier_id")
