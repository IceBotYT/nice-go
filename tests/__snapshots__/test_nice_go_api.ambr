# serializer version: 1
# name: test_get_all_barriers
  list([
    Barrier(id='test_id', type='test_type', controlLevel='test_control_level', attr=[{'key': 'test_key', 'value': 'test_value'}], state=BarrierState(deviceId='test_id', desired={'test': 'value'}, reported={'test': 'value'}, timestamp='1234567890', version=1, connectionState=ConnectionState(connected=True, updatedTimestamp='1234567890')), api=None),
  ])
# ---
# name: test_get_all_barriers_no_connection_state
  list([
    Barrier(id='test_id', type='test_type', controlLevel='test_control_level', attr=[{'key': 'test_key', 'value': 'test_value'}], state=BarrierState(deviceId='test_id', desired={'test': 'value'}, reported={'test': 'value'}, timestamp='1234567890', version=1, connectionState=None), api=None),
  ])
# ---
